discretize/unstructured.py

```diff
@@ -0,0 +1,978 @@
+import numpy as np
+import scipy.sparse as sp
+from scipy.spatial import KDTree
+from discretize.utils import Identity, invert_blocks, spzeros
+from discretize.base import BaseMesh
+from discretize._extensions.simplex_helpers import (
+    _build_faces_edges,
+    _build_adjacency,
+    _directed_search,
+    _interp_cc
+)
+from discretize.mixins import InterfaceMixins, SimplexMeshIO
+
+
+class SimplexMesh(BaseMesh, SimplexMeshIO, InterfaceMixins):
+    """Class for traingular (2D) and tetrahedral (3D) meshes.
+
+    Simplex is the abstract term for triangular like elements in an arbitrary dimension.
+    Simplex meshes are subdivided into trianglular (in 2D) or tetrahedral (in 3D)
+    elements. They are capable of representing abstract geometric surfaces, with widely
+    variable element sizes.
+
+    Parameters
+    ----------
+    nodes : (n_nodes, dim) array_like of float
+        Defines every node of the mesh.
+
+    simplices : (n_cells, dim+1) array_like of int
+        This array defines the connectivity of nodes to form cells. Each element
+        indexes into the `nodes` array. Each row defines which nodes make a given cell.
+        This array is sorted along each row and then stored on the mesh.
+
+    Notes
+    -----
+    Only rudimentary checking of the input nodes and simplices are performed, which
+    only checks for degenerate simplices who have zero volume. There are no checks for
+    overlapping cells, or for the quality of the mesh.
+
+    Examples
+    --------
+    Here we generate a basic 2D triangular mesh, by triangulating a rectangular domain.
+
+    >>> from discretize import SimplexMesh
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+    >>> import matplotlib.tri as tri
+
+    First we define the nodes of our mesh
+
+    >>> X, Y = np.mgrid[-20:20:21j, -10:10:11j]
+    >>> nodes = np.c_[X.reshape(-1), Y.reshape(-1)]
+
+    Then we triangulate the nodes, here we use matplotlib, but you could also use
+    scipy's Delaunay, or any other triangular mesh generator. Essentailly we are
+    creating every node in the mesh, and a list of triangles/tetrahedrals defining which
+    nodes make of each cell.
+
+    >>> triang = tri.Triangulation(nodes[:, 0], nodes[:, 1])
+    >>> simplices = triang.triangles
+
+    Finally we can assemble them into a SimplexMesh
+
+    >>> mesh = SimplexMesh(nodes, simplices)
+    >>> mesh.plot_grid()
+    >>> plt.show()
+    """
+    _meshType = "simplex"
+    _items = {"nodes", "simplices"}
+
+    def __init__(self, nodes, simplices):
+        # grab copies of the nodes and simplices for protection
+        nodes = np.asarray(nodes)
+        simplices = np.asarray(simplices)
+        dim = nodes.shape[1]
+        if dim not in [3, 2]:
```

Comment:
This comment is prompted by some of the conversations we had at the simpeg meeting the other week -- what are your thoughts on separating the checks into a small method `_check_dim` or similar. This would make the init code a bit simpler to follow and later if we wanted to add more complex validation, that code could expand without the init getting overly complicated